### PR TITLE
Add TSERV_GROUP_NAME, ability to balance within tserver group

### PR DESCRIPTION
**Pull request by dlmarion**
Added TSERV_GROUP_NAME property and refactored how groups are determined within the server processes. Modified LiveTServerSet to pull the group from the ServiceLockData and add it to the TServerInstance. Modified the TabletGroupWatcher to use the group information in the TServerInstance to determine if the tablet is currently hosted on a TServer that is not in the correct group

### CODE REVIEW

**Comment by dlmarion** [comment]
This replaces PR #3482 . It does not put the tserver group name into the metadata. There is likely more work to be done here w/r/t the balancers and tests. Putting up for early feedback on approach.

**Comment by dlmarion** [comment]
Follow-on tasks include:

Rename `compactor.queue` to `compactor.group`
Update scripts for `compactor.queue` name change, addition of tserver.group
TabletServer resource group tests, to include balancers

**Comment by dlmarion** [comment]
Follow-on tasks completed in https://github.com/dlmarion/accumulo/tree/3496-followon-rename-compactor-queue. Will create PR for elasticity branch once this merged.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:50]: If using the table.custom prop these docs could have text like : 

```
Set table.custom.assignment.group=<group name> to limit assignment of tablets to tablet servers that were started with the named group.   
```

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:172]: It using the custom table prop, could do the following.

```suggestion
          environment.getConfiguration(e.getKey()).getTableCustom("assignment.group");
```

File: server/base/src/main/java/org/apache/accumulo/server/manager/LiveTServerSet.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/manager/LiveTServerSet.java:408]: Its possible that depending on the uses patterns that it may be more efficient to compute an immutable map of this info when it changes.  The tserver does this for online tablets in https://github.com/apache/accumulo/blob/main/server/tserver/src/main/java/org/apache/accumulo/tserver/OnlineTablets.java

I don't know if this would be useful w/o analyzing the larger way in which this is used.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:78]: This would be a follow on, we may want to introduce a type for the group names instead of using string.  This would be something like TableId.

```suggestion
    Map<String,Set<TabletServerId>> currentResourceGroups();
```

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:130]: I don't think we should throw an exception when there are zero tablet servers, this could be temporary situation that happens normally.  Maybe log a warning and do nothing.

Also I don't think we should fall back to some default when a user configured group has no tservers in it, could log a warning and take no action.

**Review by keith-turner — APPROVED** [review]
This is really neat.  I made comments, those could be addressed as follow on issues.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:130]: It would be good to have ITs that test zero tservers in a group case.  Maybe something like the following.

 1. Create a table that requires its tablet be assigned to tservers in group X. Tserver group X currently has not tservers.
 2. Start a thread that tries to write to the above table.
 3. Start a tserver in group X.
 4. Verify the write in the thread was successful.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: it's also used in TabletGroupWatcher.getAssignmentsFromBalancer to print a warning if a tablet is assigned to a tserver outside of its resource group.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: > it's also used in TabletGroupWatcher.getAssignmentsFromBalancer to print a warning if a tablet is assigned to a tserver outside of its resource group.

I don't TGW should try to validate if a plugin is behaving properly.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: If we scope the config to plugin, it could allow TableLoadBalancer to have more complex configuration that is tightly coupled to it.  For example TableLoadBalancer could support the property `table.custom.assignment.group` or `table.custom.assignment.rangedGroups` where rangedGroups allows a user to set a json value that specifies something like the folllowing.

 *  For tablets in range `(<row1>,<row2>]` user tablet server group X
 *  For tablets in range `(<row3>,<row4>]` user tablet server group Y
 * For all other tablet use tablet server group Z

So one property or the other could be set, if both are set it throws an error.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:

https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: > So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:
>
> https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325

For that case maybe we need to add a new method to the balancer so we can ask `hey balancer are you happy with the group this tablet is currently assigned to?`

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: >  So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:
> 
> https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325


Does this check also need to run in the tablet mgmt iterator?

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: > For that case maybe we need to add a new method to the balancer so we can ask hey balancer are you happy with the group this tablet is currently assigned to?

Wondering the balancers existing `long balance(BalanceParameters params);` method can handle this case.  Can it efficiently detect tablets assigned to an undesired group and initiate migrations?

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: I don't know, I think I put the TODO there to have a conversation about it. I'm going to take another look at this before I merge.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: Yeah it needs investigation, need to figure out how to defer to the balancer plugin for this and if that needs new methods in the balancer.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: I was thinking a new interface that TableLoadBalancer implements and HostRegexTableLoadBalancer does not. Manager could check and invoke the methods on it.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: > Does this check also need to run in the tablet mgmt iterator?

I don't think it can unless we pass the group information to the iterator. The group is not in the metadata anymore, only in ZK. The LiveTServerSet object in the Manager is the only thing that knows about the TServer groups

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: If needed we could  pass the tserver group information to the tablet mgmt iterator, we already pass the live tserver set I think so this would be a similar amount of data.  We could call the balancer plugins in the tablet mgmt iterator fi needed also.  The compaction code is calling compaction plugins in CompactionJobGenerator inside tablet mgtm iterator to make decisions.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1078]: It may be good to have an IT that does something like the following.

 * Configures table A to be assigned to tserver group Z
 * Verifies assignments are as expected
 * Changes config such that table A is assigned to tserver group X
 * Starts a tserver in group X
 * Waits for tablets to be assigned to group X

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:76]: The suggestion looks the same to me, but I think I understand what you are saying.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:233]: Remove

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:76]: I think I meant to put some type name in the suggestion.  Must have gotten distracted by something while writing the comment.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java]: removed

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:141]: I added this test in db29813

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: I removed the TABLE_ASSIGNMENT_GROUP property from Property and kept it in TableLoadBalancer in the commits today. There is a lot in this thread. I need to re-read it and make sure I got to everything.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: I added a test that you suggested above. It uncovered some other issues with tables being deleted and how the TabletManagementIterator handles that. I think I have everything in this thread done.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java]: I moved the new property into TableLoadBalancer, and it's named `table.custom.assignment.group`. I updated the javadoc.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:146]: This method should be passed a tablet.  Also conf is probably not needed, that is passed into init and can be kept by the plugin if desired.

Instead of asking if something is in the correct resource group, could we instead ask the balancer what a tablets desired resource group is?

```suggestion
  /**
   * @return the name of the tablet server resource group the balancer would like to see this tablet in
   * @since 4.0.0
   */
  default String getDesiredResourceGroup(TabletId tablet) {
    return default constant;
  }
```

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/balancer/TableLoadBalancer.java:173]: This could be a follow on.  Need to determine if we need to do anything to avoid recomputing this map on every call to getAssignments.  Not sure if it matters, depends on how frequently the manager calls this.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:261]: Instead of sleeping for a fixed time, could refactor the test to wait for things to be in the expected groups.  May make the test a bit more stable.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: Some more test that could be useful.

 * Test balancing the root and metadata table to diff tserver groups.
 * Test two user tables that balance to diff groups to make sure that works as expected.
 * Test tablets within in a single table that balance to different groups.

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:146]: IIRC you suggested that we perform this check in TabletManagementIterator (TMI). I can't call `balancer.init()`from the iterator because I don't have access to `ServerContext` to construct a `BalancerEnvironmentImpl`. If someone can point me to a way around this, then I should be able to simplify the method API. I don't think `getDesiredResourceGroup` is a good name as one of the earlier suggestions was to *not* fall back to using the `default` group if the specified group did not have any servers.

This method is called from `TabletMetadata.getTabletState` as it's common to both TGW and TMI, and it seems appropriate because it returns `TabletState.ASSIGNED_TO_WRONG_GROUP`.

I'm not sure how to implement your suggestion.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: > Test two user tables that balance to diff groups to make sure that works as expected.

`testBalancerWithResourceGroups` does this. I creates one table in the default group and one table in the GROUP1 group. Then it validates that all of the first tables tablets are located in the tserver in the default group and that all of the second tables tablets are located in the tserver in the GROUP1 group.

Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: > Test balancing the root and metadata table to diff tserver groups.

Is this any different than the `testBalancerWithResourceGroups` test? Are these tables handled differently?

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java]: Removed sleep in d0b0c4d

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: The root tablet is handled differently, metadata table is more likely to be like everything else. When I added an IT for compaction of root and metadata, I found the root tablet did not work.  Had to add this :

https://github.com/apache/accumulo/blob/406552e46df4412cf8f4f52962446773a9eabee5/server/base/src/main/java/org/apache/accumulo/server/manager/state/ZooTabletStateStore.java#L79-L84

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java:146]: > I can't call balancer.init()from the iterator because I don't have access to ServerContext to construct a BalancerEnvironmentImpl

Could add a constructor like the following.  I think this is ok as this is internal Accumulo code and we know the actual type behind the ServiceEnv interface.

```java
public BalancerEnvironmentImpl(ServiceEnvironment env) {
    super(((ServiceEnvironmentImpl) env).getContext());
  }
```

> I don't think getDesiredResourceGroup is a good name as one of the earlier suggestions was to not fall back to using the default group if the specified group did not have any servers.

Thats a good point, maybe `getRequiredResourceGroup` is a better name.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: We may be able to have the root tablet use the TabletMgmtIterator, but not sure.   Going to open an issue.  It would be nice if it could, then it would be more likely to have consistent behavior.

Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java:67]: Created #3587

File: core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java]: I was able to implement your suggestions in a03bfb6. I modified the TabletBalancer api method to just take a TabletId and I renamed it to `getResourceGroup`.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java]: @keith-turner - I added balance tests for root and metadata table in 6ca1a88, and they work.

Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java]: > If the SPI supports the functionality, then it would be good to ensure it actually works. 

So, the `TabletBalancer.getResourceGroup` api method works at the tablet level now. The TabletMetadata.getTabletState method returns `TabletState.ASSIGNED_TO_WRONG_GROUP` when a tablet is not hosted on a tablet server that is in the same group that `TabletBalancer.getResourceGroup` returns. `TableLoadBalancer` passes the set of TabletServers that correspond to the resource group to the tables load balancer for assignment and balance purposes. This allows, for example, the SimpleLoadBalancer to assign and balance within the subset of tablet servers.

I think to do what you are asking (assign and balance at the tablet level based on resource group assignment), the table level balancers would need to be modified to make decisions on a per-tablet basis.

Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java]: I guess one way to resolve this is to remove `TabletBalancer.getResourceGroup` such that *only* `TableLoadBalancer` has this functionality. `TabletMetadata.getTabletState` would only check the resource group if the balancer is `TableLoadBalancer`.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:453]: his warn logging could get really spammy. Also for this case it seems like it should return TabletState.ASSIGNED_TO_WRONG_GROUP. The tserver would be unassigned and then left unassigned until some tservers showed up in the group.

**Review by keith-turner — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java]: Resolved in 899cb4e

File: test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java]: Discussed the above with @keith-turner , I'm going to merge as-is and create a follow-on issue.

**Pull request merged**
